src/Microsoft.AspNetCore.Blazor.Server/Bots/RenderBatchWriter.cs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.AspNetCore.Blazor.RenderTree;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Blazor.Server
+{
+    // TODO: We should consider *not* having this type of infrastructure in the .Server
+    // project, but instead in some new project called .Remote or similar, since it
+    // would also be used in Electron and possibly WebWorker cases.
+
+    /// <summary>
+    /// Provides a custom binary serializer for <see cref="RenderBatch"/> instances.
+    /// This is designed with both server-side and client-side perf in mind:
+    /// 
+    ///  * Array-like regions always have a fixed size per entry (even if some entry types
+    ///    don't require as much space as others) so the recipient can index directly.
+    ///  * The indices describing where field data starts, where each string value starts,
+    ///    etc., are written *after* that data, so when writing the data we don't have to
+    ///    compute the locations up front or seek back to an earlier point in the stream.
+    ///    The recipient can only process the data after reading it all into a buffer,
+    ///    so it's no disadvantage for the location info to be at the end.
+    ///  * We only serialize the data that the JS side will need. For example, we don't
+    ///    emit frame sequence numbers, or any representation of nonstring attribute
+    ///    values, or component instances, etc.
+    ///    
+    /// We don't have or need a .NET reader for this format. We only read it from JS code.
+    /// </summary>
+    internal class RenderBatchWriter : IDisposable
+    {
+        private readonly List<string> _strings;
```

Comment:
I did consider that, but since I think we'll be very likely to want to put compression around this anyway, that will have exactly the same effect (well, it will do better, because it will de-dupe common substrings as well as complete ones). So assuming we do that, having extra logic here to do poor-person's compression might be undesirable.

If later it turns out we aren't going to compress the websocket traffic then yes I definitely think we should do that.